services/py-api/src/service/hackathon_service.py

```diff
@@ -441,3 +443,85 @@ def send_successful_registration_email(
+        for participant in result.ok_value:
+            if participant.programming_level == "I am not participating as a programmer":
+                non_programming_oriented.append(participant)
+            else:
+                programming_oriented.append(participant)
+
+        return Ok((programming_oriented, non_programming_oriented))
+
+    def form_random_participant_teams(
+        self, prog_participants: list[Participant], non_prog_participants: list[Participant]
+    ) -> List[RandomTeam]:
+        # Calculate the number of hackathon participants
+        number_of_random_participants = len(prog_participants) + len(non_prog_participants)
+        # Calculate the number of teams that are going to be needed for the given number of participants
+        number_of_teams = ceil(number_of_random_participants / self.MAX_NUMBER_OF_TEAM_MEMBERS)
+
+        # Create a dictionary that is going to hold the participants for each team
+        teams = []
+        # Populate the dictionary with the Random Teams named `RandomTeam{i}`
+        for i in range(number_of_teams):
+            teams.append(RandomTeam(team=Team(name=f"RandomTeam{i}", is_verified=True), participants=[]))
+
+        # Spread all the programming experienced participants to the teams in a Round Robin (RR) manner
+        ctr = 0  # initialize a counter
+        while len(prog_participants) > 0:
+            teams[ctr % number_of_teams]["participants"].append(prog_participants.pop())
+            ctr += 1
+
+        # Spread all the non-programming experienced participants to the teams in a Round Robin (RR) manner
+        ctr = 0  # reset counter
+        while len(non_prog_participants) > 0:
+            teams[ctr % number_of_teams]["participants"].append(non_prog_participants.pop())
+            ctr += 1
+
+        return teams
+
+    async def _create_random_participant_teams_in_transaction_callback(
+        self, random_teams: List[RandomTeam], session: Optional[AsyncIOMotorClientSession] = None
+    ) -> Result[List[RandomTeam], DuplicateTeamNameError | ParticipantNotFoundError | Exception]:
+        # Loop through each tuple in the list. Create the team. Update all the participants with the id of the team
+        # created
+        for random_team in random_teams:
+            # Create the team
+            result = await self._team_repo.create(team=random_team["team"], session=session)
+
+            if is_err(result):
+                return result
+
+            # Insert all the participants in the team that was created
+            for participant in random_team["participants"]:
+
+                result = await self._participant_repo.update(
+                    obj_id=str(participant.id),
+                    obj_fields=UpdateParticipantParams(team_id=random_team["team"].id),
+                    session=session,
+                )
+
+                if is_err(result):
+                    return result
```

Comment:
nice for coming out with the RandomTeam dict for storing info there before doing the final db operation